api/v1beta1/disruption_types.go

```diff
@@ -91,13 +94,85 @@ func init() {
 }
 
 // Hash returns the disruption spec JSON hash
-func (ds *DisruptionSpec) Hash() (string, error) {
+func (s *DisruptionSpec) Hash() (string, error) {
 	// serialize instance spec to JSON
-	specBytes, err := json.Marshal(ds)
+	specBytes, err := json.Marshal(s)
 	if err != nil {
 		return "", fmt.Errorf("error serializing instance spec: %w", err)
 	}
 
 	// compute bytes hash
 	return fmt.Sprintf("%x", md5.Sum(specBytes)), nil
 }
+
+func (s *DisruptionSpec) Validate() error {
+	// Rule: no targeted container if disruption is node-level
+	if len(s.Containers) > 0 && s.Level != chaostypes.DisruptionLevelPod {
+		return errors.New("cannot target specific container of a node-level disruption")
+	}
+
+	// Rule: at least one disruption field
+	if s.DNS == nil &&
+		s.CPUPressure == nil &&
+		s.Network == nil &&
+		s.NodeFailure == nil &&
+		s.DiskPressure == nil {
+		return errors.New("cannot apply a disruption with no target process")
+	}
+
+	//Rule: DNS and Network disruptions are incompatible
+	if s.DNS != nil && s.Network != nil {
+		return errors.New("cannot apply DNS and Network disruptions concurrently")
+	}
+
+	return nil
+}
+
+// validates rules for disruption global scope and all subsequent disruption specifications
+func (s *DisruptionSpec) ValidateDisruptionSpec() error {
+	err := s.Validate()
+	if err != nil {
+		return err
+	}
+
+	for _, kind := range chaostypes.DisruptionKinds {
+		var validator chaosapi.DisruptionValidator
+
+		disruptionExists, validator, _ := s.DisruptionKindInterfaceGenerator(kind)
+		if !disruptionExists {
+			continue
+		}
+
+		err := validator.Validate()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// validates existence and generates instances of DisruptionKind Interfaces: DisruptionValidator, DisruptionArgsGenerator
+func (s *DisruptionSpec) DisruptionKindInterfaceGenerator(kind chaostypes.DisruptionKind) (bool, chaosapi.DisruptionValidator, chaosapi.DisruptionArgsGenerator) {
+	var validator chaosapi.DisruptionValidator
+
+	var generator chaosapi.DisruptionArgsGenerator
+
+	switch kind {
+	case chaostypes.DisruptionKindNodeFailure:
+		validator, generator = s.NodeFailure, s.NodeFailure
+	case chaostypes.DisruptionKindNetworkDisruption:
+		validator, generator = s.Network, s.Network
+	case chaostypes.DisruptionKindDNSDisruption:
+		validator, generator = s.DNS, s.DNS
+	case chaostypes.DisruptionKindCPUPressure:
+		validator, generator = s.CPUPressure, s.CPUPressure
+	case chaostypes.DisruptionKindDiskPressure:
+		validator, generator = s.DiskPressure, s.DiskPressure
+	}
+
+	// ensure that the underlying disruption spec is not nil
+	disruptionExists := !reflect.ValueOf(validator).IsNil() && !reflect.ValueOf(generator).IsNil()
```

Comment:
You almost certainly don't need the `&&` here, because it's the same pointer in both situations

Reply:
That is correct and will be fixed